daemon/src/bitmex_price_feed.rs

```diff
@@ -0,0 +1,107 @@
+use crate::model::Usd;
+use anyhow::Result;
+use futures::{StreamExt, TryStreamExt};
+use rust_decimal::Decimal;
+use std::convert::TryFrom;
+use std::future::Future;
+use std::time::SystemTime;
+use tokio::sync::watch;
+use tokio_tungstenite::tungstenite;
+
+/// Connects to the BitMex price feed, returning the polling task and a watch channel that will
+/// always hold the last quote.
+pub async fn new() -> Result<(impl Future<Output = ()>, watch::Receiver<Quote>)> {
+    let (connection, _) = tokio_tungstenite::connect_async(
+        "wss://www.bitmex.com/realtime?subscribe=quoteBin1m:XBTUSD",
+    )
+    .await?;
+    let mut quotes = connection
+        .inspect_err(|e| tracing::warn!("Error on websocket stream: {}", e))
+        .map(|msg| Quote::from_message(msg?))
+        .filter_map(|result| async move { result.transpose() })
+        .boxed()
+        .fuse();
+
+    tracing::info!("Connected to BitMex realtime API");
+
+    let first_quote = quotes.select_next_some().await?;
+    let (sender, receiver) = watch::channel(first_quote);
+
+    let task = async move {
+        while let Ok(Some(quote)) = quotes.try_next().await {
+            if sender.send(quote).is_err() {
+                break; // If the receiver dies, we can exit the loop.
+            }
+        }
+
+        tracing::warn!("Failed to read quote from websocket");
+    };
+
+    Ok((task, receiver))
+}
+
+#[derive(Debug)]
+pub struct Quote {
+    pub timestamp: SystemTime,
+    pub bid: Usd,
+    pub ask: Usd,
+}
+
+impl Quote {
+    fn from_message(message: tungstenite::Message) -> Result<Option<Self>> {
+        let text_message = match message {
+            tungstenite::Message::Text(text_message) => text_message,
+            _ => anyhow::bail!("Bad message type, only text is supported"),
+        };
+
+        let table_message = match serde_json::from_str::<wire::TableMessage>(&text_message) {
+            Ok(table_message) => table_message,
+            Err(_) => return Ok(None),
+        };
+
+        let [quote] = table_message.data;
+
+        Ok(Some(Self {
+            timestamp: SystemTime::now(),
```

Comment:
Why did you go for `SystemTime::now()` and not the `timestamp` from Bitmex?

Reply:
We could take it as `String` in the serde struct and then parse it here. Do you want me to do that?

Reply:
I think it's fine. The only reason why this could be relevant is if we want to check if the rate we received is current. We don't do any math with the rate (yet) so I guess that is not important? 🤷‍♂️